File: src/fab-go/cmd/fab/panemap.go
package main

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"

	"github.com/spf13/cobra"
	"github.com/wvrdz/fab-kit/src/fab-go/internal/resolve"
	sf "github.com/wvrdz/fab-kit/src/fab-go/internal/statusfile"
	"github.com/wvrdz/fab-kit/src/fab-go/internal/status"
	"gopkg.in/yaml.v3"
)

func paneMapCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "pane-map",
		Short: "Show tmux pane-to-worktree mapping with fab pipeline state",
		Args:  cobra.NoArgs,
		RunE:  runPaneMap,
	}
}

// paneEntry holds a single tmux pane's ID, tab (window) name, and current working directory.
type paneEntry struct {
	id  string
	tab string
	cwd string
}

// paneRow holds the resolved data for a single output row.
type paneRow struct {
	pane      string
	tab       string
	worktree  string
	change    string
	stage     string
	agent     string
}

func runPaneMap(cmd *cobra.Command, args []string) error {
	// Tmux session guard
	if os.Getenv("TMUX") == "" {
		fmt.Fprintln(cmd.ErrOrStderr(), "Error: not inside a tmux session")
		os.Exit(1)
	}

	// Discover tmux panes
	panes, err := discoverPanes()
	if err != nil {
		return err
	}

	// Determine main worktree root for relative path computation.
	// We derive mainRoot from the first resolved pane's git root so that
	// the command works even when invoked from outside the repo.
	mainRoot := findMainWorktreeRoot(panes)

	// Resolve each pane to a row
	var rows []paneRow
	// Cache runtime files per worktree root to avoid re-reading
	runtimeCache := make(map[string]interface{}) // wtRoot -> loaded map or error sentinel

	for _, p := range panes {
		row, ok := resolvePane(p, mainRoot, runtimeCache)
		if ok {
			rows = append(rows, row)
		}
	}

	// Output
	if len(rows) == 0 {
		fmt.Fprintln(cmd.OutOrStdout(), "No fab worktrees found in tmux panes.")
		return nil
	}

	printPaneTable(cmd, rows)
	return nil
}

// discoverPanes runs `tmux list-panes -s` (current session only) and parses the output.
// Uses tab as the field delimiter so that window names containing spaces are handled correctly.
func discoverPanes() ([]paneEntry, error) {
	out, err := exec.Command("tmux", "list-panes", "-s", "-F", "#{pane_id}\t#{window_name}\t#{pane_current_path}").Output()
	if err != nil {
		return nil, fmt.Errorf("tmux list-panes: %w", err)
	}

	var panes []paneEntry
	for _, line := range strings.Split(strings.TrimSpace(string(out)), "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		parts := strings.SplitN(line, "\t", 3)
		if len(parts) != 3 {
			continue
		}
		panes = append(panes, paneEntry{id: parts[0], tab: parts[1], cwd: parts[2]})
	}
	return panes, nil
}

// findMainWorktreeRoot returns the main worktree root by parsing `git worktree list --porcelain`.
// It derives the root from one of the discovered pane CWDs (via `git -C`) so that
// the command works even when invoked from outside the repo.
// Returns empty string if detection fails.
func findMainWorktreeRoot(panes []paneEntry) string {
	for _, p := range panes {
		out, err := exec.Command("git", "-C", p.cwd, "worktree", "list", "--porcelain").Output()
		if err != nil {
			continue
		}
		for _, line := range strings.Split(string(out), "\n") {
			if strings.HasPrefix(line, "worktree ") {
				return strings.TrimPrefix(line, "worktree ")
			}
		}
	}
	return ""
}

// resolvePane resolves a pane entry into a table row.
// Returns false if the pane should be excluded (non-git, non-fab).
func resolvePane(p paneEntry, mainRoot string, runtimeCache map[string]interface{}) (paneRow, bool) {
	// Resolve git worktree root
	wtRoot, err := gitWorktreeRoot(p.cwd)
	if err != nil {
		return paneRow{}, false
	}

	// Check for fab/ directory
	fabDir := filepath.Join(wtRoot, "fab")
	if _, err := os.Stat(fabDir); os.IsNotExist(err) {
		return paneRow{}, false
	}

	// Compute worktree display path
	wtDisplay := worktreeDisplayPath(wtRoot, mainRoot)

	// Read .fab-status.yaml symlink for the active change folder name
	changeName, folderName := readFabCurrent(wtRoot)

	// Read stage from .status.yaml
	stageName := "\u2014" // em dash
	if folderName != "" {
		statusPath := filepath.Join(fabDir, "changes", folderName, ".status.yaml")
		if statusFile, err := sf.Load(statusPath); err == nil {
			stage, _ := status.DisplayStage(statusFile)
			stageName = stage
		}
	}

	// Determine agent state from .fab-runtime.yaml
	agentState := resolveAgentState(wtRoot, folderName, runtimeCache)

	return paneRow{
		pane:     p.id,
		tab:      p.tab,
		worktree: wtDisplay,
		change:   changeName,
		stage:    stageName,
		agent:    agentState,
	}, true
}

// gitWorktreeRoot returns the git worktree root for a given path.
func gitWorktreeRoot(dir string) (string, error) {
	cmd := exec.Command("git", "-C", dir, "rev-parse", "--show-toplevel")
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(out)), nil
}

// worktreeDisplayPath computes the display path for a worktree.
// Main worktree shows "(main)", others show path relative to main's parent.
func worktreeDisplayPath(wtRoot, mainRoot string) string {
	if mainRoot != "" && wtRoot == mainRoot {
		return "(main)"
	}
	if mainRoot != "" {
		parent := filepath.Dir(mainRoot)
		rel, err := filepath.Rel(parent, wtRoot)
		if err == nil {
			return rel + "/"
		}
	}
	// Fallback: basename with trailing slash
	return filepath.Base(wtRoot) + "/"
}

// readFabCurrent reads .fab-status.yaml symlink and returns (displayName, folderName).
// displayName is what shows in the Change column; folderName is the actual folder.
func readFabCurrent(wtRoot string) (string, string) {
	symlinkPath := filepath.Join(wtRoot, ".fab-status.yaml")
	target, err := os.Readlink(symlinkPath)
	if err != nil {
		return "(no change)", ""
	}

	// target is "fab/changes/{name}/.status.yaml"
	folderName := resolve.ExtractFolderFromSymlink(target)
	if folderName == "" {
		return "(no change)", ""
	}

	return folderName, folderName
}

// resolveAgentState determines the agent state string for a pane row.
func resolveAgentState(wtRoot, folderName string, cache map[string]interface{}) string {
	if folderName == "" {
		return "\u2014" // em dash for no change
	}

	rtPath := filepath.Join(wtRoot, ".fab-runtime.yaml")

	// Check cache
	var rtData map[string]interface{}
	var fileMissing bool

	if cached, ok := cache[wtRoot]; ok {
		switch v := cached.(type) {
		case map[string]interface{}:
			rtData = v
		case nil:
			fileMissing = true
		}
	} else {
		// Load and cache
		loaded, err := loadPaneMapRuntimeFile(rtPath)
		if err != nil {
			if os.IsNotExist(err) {
				// Cache absence of runtime file so we don't re-check repeatedly.
				cache[wtRoot] = nil
				fileMissing = true
			} else {
				// Do not cache nil for non-NotExist errors; they may be transient
				// or indicate a corrupted .fab-runtime.yaml. Surface a warning.
				fmt.Fprintf(os.Stderr, "warning: failed to load %s: %v\n", rtPath, err)
				return "?"
			}
		} else {
			cache[wtRoot] = loaded
			rtData = loaded
		}
	}

	if fileMissing {
		return "?"
	}

	// Look up change entry
	folderEntry, ok := rtData[folderName].(map[string]interface{})
	if !ok {
		return "active"
	}

	agentBlock, ok := folderEntry["agent"].(map[string]interface{})
	if !ok {
		return "active"
	}

	idleSince, ok := agentBlock["idle_since"]
	if !ok {
		return "active"
	}

	// Parse idle_since as Unix timestamp
	var ts int64
	switch v := idleSince.(type) {
	case int:
		ts = int64(v)
	case int64:
		ts = v
	case float64:
		ts = int64(v)
	default:
		return "active"
	}

	elapsed := time.Now().Unix() - ts
	if elapsed < 0 {
		elapsed = 0
	}

	return fmt.Sprintf("idle (%s)", formatIdleDuration(elapsed))
}

// loadPaneMapRuntimeFile reads and parses .fab-runtime.yaml.
// Returns an error if the file doesn't exist (distinct from loadRuntimeFile in runtime.go
// which returns an empty map for missing files — we need to distinguish missing vs empty).
func loadPaneMapRuntimeFile(path string) (map[string]interface{}, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, err // includes os.IsNotExist
	}
	var m map[string]interface{}
	if err := yaml.Unmarshal(data, &m); err != nil {
		return nil, fmt.Errorf("parsing %s: %w", path, err)
	}
	if m == nil {
		m = make(map[string]interface{})
	}
	return m, nil
}

// formatIdleDuration formats elapsed seconds into a human-readable duration.
// Uses floor division: <60s -> Ns, 60s-3599s -> Nm, >=3600s -> Nh.
func formatIdleDuration(seconds int64) string {
	if seconds < 60 {
		return fmt.Sprintf("%ds", seconds)
	}
	if seconds < 3600 {
		return fmt.Sprintf("%dm", seconds/60)
	}
	return fmt.Sprintf("%dh", seconds/3600)
}

// printPaneTable prints the aligned pane map table.
func printPaneTable(cmd *cobra.Command, rows []paneRow) {
	// Compute column widths
	headers := [6]string{"Pane", "Tab", "Worktree", "Change", "Stage", "Agent"}
	widths := [6]int{len(headers[0]), len(headers[1]), len(headers[2]), len(headers[3]), len(headers[4]), len(headers[5])}

	for _, r := range rows {
		cols := [6]string{r.pane, r.tab, r.worktree, r.change, r.stage, r.agent}
		for i, c := range cols {
			if len(c) > widths[i] {
				widths[i] = len(c)
			}
		}
	}

	// Print header
	fmtStr := fmt.Sprintf("%%-%ds  %%-%ds  %%-%ds  %%-%ds  %%-%ds  %%s\n", widths[0], widths[1], widths[2], widths[3], widths[4])
	fmt.Fprintf(cmd.OutOrStdout(), fmtStr, headers[0], headers[1], headers[2], headers[3], headers[4], headers[5])

	// Print data rows
	for _, r := range rows {
		fmt.Fprintf(cmd.OutOrStdout(), fmtStr, r.pane, r.tab, r.worktree, r.change, r.stage, r.agent)
	}
}

File: src/fab-rust/src/panemap.rs
use anyhow::{bail, Result};
use std::collections::HashMap;
use std::fs;
use std::path::{Path, PathBuf};
use std::process::Command;

use crate::resolve;
use crate::status;
use crate::statusfile;

struct PaneEntry {
    id: String,
    cwd: String,
}

struct PaneRow {
    pane: String,
    worktree: String,
    change: String,
    stage: String,
    agent: String,
}

pub fn run_pane_map() -> Result<()> {
    // Tmux session guard
    if std::env::var("TMUX").unwrap_or_default().is_empty() {
        bail!("not inside a tmux session");
    }

    // Discover tmux panes
    let panes = discover_panes()?;

    // Determine main worktree root
    let main_root = find_main_worktree_root(&panes);

    // Resolve each pane to a row
    let mut rows = Vec::new();
    let mut runtime_cache: HashMap<String, Option<HashMap<String, serde_yaml::Value>>> = HashMap::new();

    for p in &panes {
        if let Some(row) = resolve_pane(p, &main_root, &mut runtime_cache) {
            rows.push(row);
        }
    }

    // Output
    if rows.is_empty() {
        println!("No fab worktrees found in tmux panes.");
        return Ok(());
    }

    print_pane_table(&rows);
    Ok(())
}

fn discover_panes() -> Result<Vec<PaneEntry>> {
    let output = Command::new("tmux")
        .args(["list-panes", "-a", "-F", "#{pane_id} #{pane_current_path}"])
        .output()?;

    if !output.status.success() {
        bail!("tmux list-panes: {}", String::from_utf8_lossy(&output.stderr));
    }

    let stdout = String::from_utf8_lossy(&output.stdout);
    let mut panes = Vec::new();
    for line in stdout.trim().lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        let parts: Vec<&str> = line.splitn(2, ' ').collect();
        if parts.len() != 2 {
            continue;
        }
        panes.push(PaneEntry {
            id: parts[0].to_string(),
            cwd: parts[1].to_string(),
        });
    }
    Ok(panes)
}

fn find_main_worktree_root(panes: &[PaneEntry]) -> String {
    for p in panes {
        if let Ok(output) = Command::new("git")
            .args(["-C", &p.cwd, "worktree", "list", "--porcelain"])
            .output()
        {
            if output.status.success() {
                let stdout = String::from_utf8_lossy(&output.stdout);
                for line in stdout.lines() {
                    if let Some(path) = line.strip_prefix("worktree ") {
                        return path.to_string();
                    }
                }
            }
        }
    }
    String::new()
}

fn resolve_pane(
    p: &PaneEntry,
    main_root: &str,
    runtime_cache: &mut HashMap<String, Option<HashMap<String, serde_yaml::Value>>>,
) -> Option<PaneRow> {
    // Resolve git worktree root
    let wt_root = git_worktree_root(&p.cwd)?;

    // Check for fab/ directory
    let fab_dir = PathBuf::from(&wt_root).join("fab");
    if !fab_dir.exists() {
        return None;
    }

    // Compute worktree display path
    let wt_display = worktree_display_path(&wt_root, main_root);

    // Read .fab-status.yaml symlink
    let (change_name, folder_name) = read_fab_current(&wt_root);

    // Read stage from .status.yaml
    let stage_name = if !folder_name.is_empty() {
        let status_path = fab_dir
            .join("changes")
            .join(&folder_name)
            .join(".status.yaml");
        if let Ok(sf) = statusfile::load(&status_path.to_string_lossy()) {
            let (stage, _) = status::display_stage(&sf);
            stage
        } else {
            "\u{2014}".to_string() // em dash
        }
    } else {
        "\u{2014}".to_string() // em dash
    };

    // Determine agent state
    let agent_state = resolve_agent_state(&wt_root, &folder_name, runtime_cache);

    Some(PaneRow {
        pane: p.id.clone(),
        worktree: wt_display,
        change: change_name,
        stage: stage_name,
        agent: agent_state,
    })
}

fn git_worktree_root(dir: &str) -> Option<String> {
    let output = Command::new("git")
        .args(["-C", dir, "rev-parse", "--show-toplevel"])
        .output()
        .ok()?;

    if !output.status.success() {
        return None;
    }

    Some(
        String::from_utf8_lossy(&output.stdout)
            .trim()
            .to_string(),
    )
}

fn worktree_display_path(wt_root: &str, main_root: &str) -> String {
    if !main_root.is_empty() && wt_root == main_root {
        return "(main)".to_string();
    }
    if !main_root.is_empty() {
        let parent = Path::new(main_root).parent().unwrap_or(Path::new("/"));
        if let Ok(rel) = pathdiff(wt_root, &parent.to_string_lossy()) {
            return format!("{}/", rel);
        }
    }
    // Fallback: basename with trailing slash
    format!(
        "{}/",
        Path::new(wt_root)
            .file_name()
            .map(|n| n.to_string_lossy().to_string())
            .unwrap_or_default()
    )
}

fn pathdiff(path: &str, base: &str) -> Result<String, ()> {
    let path = Path::new(path);
    let base = Path::new(base);
    path.strip_prefix(base)
        .map(|p| p.to_string_lossy().to_string())
        .map_err(|_| ())
}

fn read_fab_current(wt_root: &str) -> (String, String) {
    let symlink_path = PathBuf::from(wt_root).join(".fab-status.yaml");
    match fs::read_link(&symlink_path) {
        Ok(target) => {
            let target_str = target.to_string_lossy().to_string();
            let folder_name = resolve::extract_folder_from_symlink(&target_str);
            if folder_name.is_empty() {
                ("(no change)".to_string(), String::new())
            } else {
                (folder_name.clone(), folder_name)
            }
        }
        Err(_) => ("(no change)".to_string(), String::new()),
    }
}

fn resolve_agent_state(
    wt_root: &str,
    folder_name: &str,
    cache: &mut HashMap<String, Option<HashMap<String, serde_yaml::Value>>>,
) -> String {
    if folder_name.is_empty() {
        return "\u{2014}".to_string(); // em dash
    }

    let rt_path = PathBuf::from(wt_root).join(".fab-runtime.yaml");

    // Check cache
    if !cache.contains_key(wt_root) {
        match load_panemap_runtime_file(&rt_path.to_string_lossy()) {
            Ok(data) => {
                cache.insert(wt_root.to_string(), Some(data));
            }
            Err(e) => {
                if e.kind() == std::io::ErrorKind::NotFound {
                    cache.insert(wt_root.to_string(), None);
                } else {
                    eprintln!("warning: failed to load {}: {}", rt_path.display(), e);
                    return "?".to_string();
                }
            }
        }
    }

    let rt_data = match cache.get(wt_root) {
        Some(Some(data)) => data,
        Some(None) => return "?".to_string(), // file missing
        None => return "?".to_string(),
    };

    // Look up change entry
    let folder_entry = match rt_data.get(folder_name) {
        Some(v) => v,
        None => return "active".to_string(),
    };

    let folder_map = match folder_entry.as_mapping() {
        Some(m) => m,
        None => return "active".to_string(),
    };

    let agent_block = match folder_map.get(&serde_yaml::Value::String("agent".to_string())) {
        Some(serde_yaml::Value::Mapping(m)) => m,
        _ => return "active".to_string(),
    };

    let idle_since = match agent_block.get(&serde_yaml::Value::String("idle_since".to_string())) {
        Some(v) => v,
        None => return "active".to_string(),
    };

    let ts = match idle_since {
        serde_yaml::Value::Number(n) => n.as_i64().unwrap_or(0),
        _ => return "active".to_string(),
    };

    let now = chrono::Utc::now().timestamp();
    let elapsed = (now - ts).max(0);

    format!("idle ({})", format_idle_duration(elapsed))
}

fn load_panemap_runtime_file(path: &str) -> Result<HashMap<String, serde_yaml::Value>, std::io::Error> {
    let data = fs::read_to_string(path)?;
    let v: serde_yaml::Value = serde_yaml::from_str(&data).map_err(|e| {
        std::io::Error::new(std::io::ErrorKind::InvalidData, format!("parsing {}: {}", path, e))
    })?;
    let mut result = HashMap::new();
    if let serde_yaml::Value::Mapping(m) = v {
        for (k, v) in m {
            if let serde_yaml::Value::String(key) = k {
                result.insert(key, v);
            }
        }
    }
    Ok(result)
}

fn format_idle_duration(seconds: i64) -> String {
    if seconds < 60 {
        format!("{}s", seconds)
    } else if seconds < 3600 {
        format!("{}m", seconds / 60)
    } else {
        format!("{}h", seconds / 3600)
    }
}

fn print_pane_table(rows: &[PaneRow]) {
    let headers = ["Pane", "Worktree", "Change", "Stage", "Agent"];
    let mut widths = [
        headers[0].len(),
        headers[1].len(),
        headers[2].len(),
        headers[3].len(),
        headers[4].len(),
    ];

    for r in rows {
        let cols = [&r.pane, &r.worktree, &r.change, &r.stage, &r.agent];
        for (i, c) in cols.iter().enumerate() {
            if c.len() > widths[i] {
                widths[i] = c.len();
            }
        }
    }

    // Print header
    println!(
        "{:<w0$}  {:<w1$}  {:<w2$}  {:<w3$}  {}",
        headers[0], headers[1], headers[2], headers[3], headers[4],
        w0 = widths[0], w1 = widths[1], w2 = widths[2], w3 = widths[3],
    );

    // Print data rows
    for r in rows {
        println!(
            "{:<w0$}  {:<w1$}  {:<w2$}  {:<w3$}  {}",
            r.pane, r.worktree, r.change, r.stage, r.agent,
            w0 = widths[0], w1 = widths[1], w2 = widths[2], w3 = widths[3],
        );
    }
}
